### PR TITLE
Admiral 42 collisions

### PR DESCRIPTION
**Pull request by rusu24edward**
@glatt1 Initial implementation of a collision actor. Please take a look and let me know if I've missed anything when I moved them over. The behavior of the particles seems a little weird to me, so I just want to check that I got the implementation correct.

I'm a bit confused about the purpose of `undo_overlap`. Is it trying to capture collisions that happened sometime between the two time-steps, and then move the agent positions back to that location?

### CODE REVIEW

**Comment by glatt1** [comment]
> I'm a bit confused about the purpose of `undo_overlap`. Is it trying to capture collisions that happened sometime between the two time-steps, and then move the agent positions back to that location?

so, because we have basically discrete movements (location update at every timestep) it may be that two agents overlap physically when a collision is detected. In that case we move them both a bit back on their original trajectory so that they just touch. Otherwise fast agents might be on top of other agents from time to time which should not be allowed.

File: admiral/envs/components/examples/simple_particle.py [hunk 1]
```diff
@@ -36,7 +38,10 @@ def reset(self, **kwargs):
     def step(self, action_dict, **kwargs):
         for agent, action in action_dict.items():
             self.move_actor.process_move(self.agents[agent], action.get("accelerate", np.zeros(2)), **kwargs)
-            self.velocity_state.apply_friction(self.agents[agent])
+            self.velocity_state.apply_friction(self.agents[agent], **kwargs)
+        
+        self.render(**kwargs)
+        self.collision_actor.detect_collisions_and_modify_states(**kwargs)
```
Comment (glatt1) [inline admiral/envs/components/examples/simple_particle.py:44]: this should come before the render (or it would show overlapping agents)

**Comment by rusu24edward** [comment]
> so, because we have basically discrete movements (location update at every timestep) it may be that two agents overlap physically when a collision is detected. In that case we move them both a bit back on their original trajectory so that they just touch. Otherwise fast agents might be on top of other agents from time to time which should not be allowed because that might trigger a new collision at the next step and so the move directions would be changed continuously "backing" the agents together.

Thanks for this explanation. I'm curious if the implementation is correct? I was playing with some scenarios, and I saw two agents collide. The `update_position` actually moved them closer to each other so that they overlapped even more. I believe this happened because the collision was calculated on the "tail end" of their collision (i.e. they already moved passed the point of being closest to each other, but were still close enough to trigger a collision). `update_position` moved them back along their trajectories a bit, but it didn't seem to move them far enough because they just moved closer to each other. Is this consistent with what you would expect to see?

File: admiral/envs/components/examples/simple_particle.py [hunk 2]
```diff
@@ -2,14 +2,14 @@
...
+from admiral.envs.components.agent import VelocityAgent, PositionAgent, MassAgent, SizeAgent
 from admiral.envs.components.state import VelocityState, ContinuousPositionState
-from admiral.envs.components.actor import AccelerationMovementActor
+from admiral.envs.components.actor import AccelerationMovementActor, ContinuousCollisionActor
 from admiral.envs.components.observer import VelocityObserver, PositionObserver
 from admiral.envs import AgentBasedSimulation
 from admiral.tools.matplotlib_utils import mscatter
 
-class ParticleAgent(VelocityAgent, PositionAgent): pass
+class ParticleAgent(VelocityAgent, PositionAgent, MassAgent, SizeAgent): pass
```
Comment (glatt1) [inline admiral/envs/components/examples/simple_particle.py:12]: The movement of the agents is very uniform in this setting. Once I leave out the MassAgent the movement becomes random as expected. I have played with this and made a single agent that I just called PhysicsAgent with mass and size and it was random as well. Weird.

**Comment by glatt1** [comment]
> Thanks for this explanation. I'm curious if the implementation is correct? I was playing with some scenarios, and I saw two agents collide. The `update_position` actually moved them closer to each other so that they overlapped even more. I believe this happened because the collision was calculated on the "tail end" of their collision (i.e. they already moved passed the point of being closest to each other, but were still close enough to trigger a collision). `update_position` moved them back along their trajectories a bit, but it didn't seem to move them far enough because they just moved closer to each other. Is this consistent with what you would expect to see?

No, the agents should bounce off of each other without ever visually overlapping. I have changed the example a bit to better see what is going on but it just seems to ignore the overlap. I'll look a bit closer now why this is happening.
`
agents = {f'agent{i}': ParticleAgent(
        id=f'agent{i}',
        max_speed=.1,
        max_acceleration=.1,
        mass=1,
        size=3.,
    ) for i in range(20)}

    env = ParticleEnv(
        agents=agents,
        region=10,
        friction=0.0
    )
    fig = plt.figure()
    env.reset()
    env.render(fig=fig)

    for _ in range(100):
        env.step({agent.id: agent.action_space.sample() for agent in agents.values()}, fig=fig)
        env.render(fig=fig)
`

**Comment by glatt1** [comment]
This might be a problem. I added the size to the render function in simple_particle.py line 60. The size needs a scaling factor of 100, so that might be the reason why the agents are not colliding because they really don't on the original scale?!

`
agents_size = [100*agent.size for agent in self.agents.values()]
mscatter(agents_x, agents_y, ax=ax, m='o', s=agents_size, edgecolor='black', facecolor='gray')
`

**Comment by rusu24edward** [comment]
> so that might be the reason why the agents are not colliding

Ya, I didn't use the visualization to determine colliding, but I just dropped breakpoints in the collision function and stepped through it. That's why you see the random `render` call in the `step` function. Correlating the visual size of the agent with the actual size will be nice for viz.

**Comment by rusu24edward** [comment]
- [ ] Reset positions should not overlap agents
- [ ] Rendering factor to calculate correct rendering size
- [ ] Duplicate scenarios from the two codes and check for the same behavior.

**Comment by rusu24edward** [comment]
@glatt1 I believe this PR is ready for a full review now. I've resolved the issues we talked about and created extra tickets where needed. The only thing I haven't done here is duplicate your scenarios here. I can try that, but it would be a bit easier with your help since I have some trouble understanding the important pieces of the scenarios you have.

**Comment by rusu24edward** [comment]
> The only thing I haven't done here is duplicate your scenarios here. I can try that, but it would be a bit easier with your help since I have some trouble understanding the important pieces of the scenarios you have.

I've decided to add the scenarios after the features since many scenarios depend on multiple features, and it would be better to have all the features in place first and then test them out.

**Pull request merged**